File: packages/shared/src/types/ipd.ts
/** Ward type options. */
export type WardType =
  | "general"
  | "icu"
  | "hdu"
  | "maternity"
  | "paediatric"
  | "nicu"
  | "psychiatric"
  | "isolation"
  | "surgical"
  | "burns";

/** Bed status options. */
export type BedStatus =
  | "available"
  | "occupied"
  | "reserved"
  | "maintenance"
  | "cleaning";

/** Admission status lifecycle. */
export type AdmissionStatus =
  | "admitted"
  | "on_leave"
  | "transferred"
  | "discharged"
  | "deceased";

/** Discharge type options. */
export type DischargeType =
  | "improved"
  | "recovered"
  | "dama"
  | "absconded"
  | "deceased"
  | "transferred"
  | "referred";

/** Nursing note type. */
export type NursingNoteType =
  | "observation"
  | "medication"
  | "procedure"
  | "shift_handover"
  | "intake_output"
  | "pain_assessment"
  | "wound_care";

// ── Ward ─────────────────────────────────────────────────────────────────────

/** Ward response from the API. */
export interface WardResponse {
  id: string;
  name: string;
  code: string;
  ward_type: WardType;
  department_id: string | null;
  floor: string | null;
  total_beds: number;
  gender_restriction: string | null;
  charge_per_day_cents: number;
  is_active: boolean;
  notes: string | null;
  available_beds: number;
  occupied_beds: number;
}

/** Payload for creating a ward. */
export interface WardCreate {
  name: string;
  code: string;
  ward_type: WardType;
  department_id?: string | null;
  floor?: string | null;
  total_beds?: number;
  gender_restriction?: string | null;
  charge_per_day_cents?: number;
  notes?: string | null;
}

// ── Bed ──────────────────────────────────────────────────────────────────────

/** Bed response from the API. */
export interface BedResponse {
  id: string;
  ward_id: string;
  bed_number: string;
  bed_type: string;
  status: BedStatus;
  current_patient_id: string | null;
  current_admission_id: string | null;
  notes: string | null;
  ward_name: string | null;
  patient_name: string | null;
}

/** Payload for creating a bed. */
export interface BedCreate {
  ward_id: string;
  bed_number: string;
  bed_type?: string;
  notes?: string | null;
}

// ── Admission ────────────────────────────────────────────────────────────────

/** Admission response from the API. */
export interface AdmissionResponse {
  id: string;
  encounter_id: string;
  patient_id: string;
  ward_id: string;
  bed_id: string;
  admission_number: string;
  attending_doctor_id: string | null;
  primary_nurse_id: string | null;
  status: AdmissionStatus;
  admission_reason: string | null;
  admission_diagnosis: string | null;
  admitted_from: string | null;
  accommodation_type: string | null;
  admitted_at: string;
  discharged_at: string | null;
  discharge_type: DischargeType | null;
  discharge_diagnosis: string | null;
  length_of_stay_days: number | null;
  created_at: string;
}

/** Admission list item with patient/ward info. */
export interface AdmissionListItem {
  id: string;
  admission_number: string;
  encounter_id: string;
  patient_id: string;
  patient_name: string | null;
  patient_mrn: string | null;
  ward_name: string | null;
  bed_number: string | null;
  status: AdmissionStatus;
  admission_diagnosis: string | null;
  admitted_from: string | null;
  admitted_at: string;
  length_of_stay_days: number | null;
}

/** Admission list response. */
export interface AdmissionListResponse {
  items: AdmissionListItem[];
  total: number;
}

/** Payload for admitting a patient. */
export interface AdmissionCreate {
  encounter_id: string;
  patient_id: string;
  ward_id: string;
  bed_id: string;
  attending_doctor_id?: string | null;
  primary_nurse_id?: string | null;
  admission_reason?: string | null;
  admission_diagnosis?: string | null;
  admitted_from?: string | null;
  accommodation_type?: string | null;
}

/** Payload for discharging a patient. */
export interface DischargeRequest {
  discharge_type: DischargeType;
  discharge_diagnosis?: string | null;
  discharge_summary?: string | null;
  follow_up_plan?: string | null;
  discharge_medications?: string | null;
}

// ── Nursing Note ─────────────────────────────────────────────────────────────

/** Nursing note response. */
export interface NursingNoteResponse {
  id: string;
  admission_id: string;
  patient_id: string;
  author_id: string;
  note_type: NursingNoteType;
  content: string;
  shift: string | null;
  severity: string | null;
  created_at: string;
}

/** Payload for creating a nursing note. */
export interface NursingNoteCreate {
  note_type: NursingNoteType;
  content: string;
  shift?: string | null;
  severity?: string | null;
}

// ── Dashboard ────────────────────────────────────────────────────────────────

/** Ward board summary for IPD dashboard. */
export interface WardBoardSummary {
  total_wards: number;
  total_beds: number;
  occupied_beds: number;
  available_beds: number;
  active_admissions: number;
  occupancy_rate: number;
}

File: services/api-gateway/app/schemas/ipd.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


# ── Ward schemas ──────────────────────────────────────────────────────────────


class WardCreate(BaseModel):
    """Schema for creating a ward."""

    name: str = Field(..., min_length=1, max_length=100)
    code: str = Field(..., min_length=1, max_length=20)
    ward_type: str = Field(
        ...,
        pattern=r"^(general|icu|hdu|maternity|paediatric|nicu|psychiatric|isolation|surgical|burns)$",
    )
    department_id: uuid.UUID | None = None
    floor: str | None = Field(None, max_length=20)
    total_beds: int = Field(default=0, ge=0)
    gender_restriction: str | None = Field(
        None, pattern=r"^(male|female|any)$"
    )
    charge_per_day_cents: int = Field(default=0, ge=0)
    notes: str | None = Field(None, max_length=1000)


class WardResponse(BaseModel):
    """Schema for ward API responses."""

    id: uuid.UUID
    name: str
    code: str
    ward_type: str
    department_id: uuid.UUID | None
    floor: str | None
    total_beds: int
    gender_restriction: str | None
    charge_per_day_cents: int
    is_active: bool
    notes: str | None
    available_beds: int = 0
    occupied_beds: int = 0

    model_config = {"from_attributes": True}


# ── Bed schemas ───────────────────────────────────────────────────────────────


class BedCreate(BaseModel):
    """Schema for creating a bed."""

    ward_id: uuid.UUID
    bed_number: str = Field(..., min_length=1, max_length=20)
    bed_type: str = Field(
        default="standard",
        pattern=r"^(standard|icu|hdu|cot|incubator|isolation)$",
    )
    notes: str | None = Field(None, max_length=500)


class BedResponse(BaseModel):
    """Schema for bed API responses."""

    id: uuid.UUID
    ward_id: uuid.UUID
    bed_number: str
    bed_type: str
    status: str
    current_patient_id: uuid.UUID | None
    current_admission_id: uuid.UUID | None
    notes: str | None
    ward_name: str | None = None
    patient_name: str | None = None

    model_config = {"from_attributes": True}


# ── Admission schemas ─────────────────────────────────────────────────────────


class AdmissionCreate(BaseModel):
    """Schema for admitting a patient."""

    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    ward_id: uuid.UUID
    bed_id: uuid.UUID
    attending_doctor_id: uuid.UUID | None = None
    primary_nurse_id: uuid.UUID | None = None
    admission_reason: str | None = Field(None, max_length=2000)
    admission_diagnosis: str | None = Field(None, max_length=500)
    admitted_from: str | None = Field(
        None, pattern=r"^(opd|emergency|referral|direct)$"
    )
    accommodation_type: str | None = Field(
        None, pattern=r"^(general|semi_private|private|icu|hdu)$"
    )


class AdmissionResponse(BaseModel):
    """Schema for admission API responses."""

    id: uuid.UUID
    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    ward_id: uuid.UUID
    bed_id: uuid.UUID
    admission_number: str
    attending_doctor_id: uuid.UUID | None
    primary_nurse_id: uuid.UUID | None
    status: str
    admission_reason: str | None
    admission_diagnosis: str | None
    admitted_from: str | None
    accommodation_type: str | None
    admitted_at: datetime
    discharged_at: datetime | None
    discharge_type: str | None
    discharge_diagnosis: str | None
    length_of_stay_days: int | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AdmissionListItem(BaseModel):
    """Admission item for ward board / list views."""

    id: uuid.UUID
    admission_number: str
    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    patient_name: str | None = None
    patient_mrn: str | None = None
    ward_name: str | None = None
    bed_number: str | None = None
    status: str
    admission_diagnosis: str | None
    admitted_from: str | None
    admitted_at: datetime
    length_of_stay_days: int | None = None

    model_config = {"from_attributes": True}


class AdmissionListResponse(BaseModel):
    """Paginated admission list."""

    items: list[AdmissionListItem]
    total: int


# ── Discharge schemas ─────────────────────────────────────────────────────────


class DischargeRequest(BaseModel):
    """Schema for discharging a patient."""

    discharge_type: str = Field(
        ...,
        pattern=r"^(improved|recovered|dama|absconded|deceased|transferred|referred)$",
    )
    discharge_diagnosis: str | None = Field(None, max_length=500)
    discharge_summary: str | None = Field(None, max_length=5000)
    follow_up_plan: str | None = Field(None, max_length=2000)
    discharge_medications: str | None = Field(None, max_length=2000)


# ── Nursing Note schemas ─────────────────────────────────────────────────────


class NursingNoteCreate(BaseModel):
    """Schema for creating a nursing note."""

    note_type: str = Field(
        ...,
        pattern=r"^(observation|medication|procedure|shift_handover|intake_output|pain_assessment|wound_care)$",
    )
    content: str = Field(..., min_length=1, max_length=5000)
    shift: str | None = Field(None, pattern=r"^(morning|afternoon|night)$")
    severity: str | None = Field(None, pattern=r"^(normal|warning|critical)$")


class NursingNoteResponse(BaseModel):
    """Schema for nursing note API responses."""

    id: uuid.UUID
    admission_id: uuid.UUID
    patient_id: uuid.UUID
    author_id: uuid.UUID
    note_type: str
    content: str
    shift: str | None
    severity: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Ward Board (bed occupancy overview) ───────────────────────────────────────


class WardBoardSummary(BaseModel):
    """Ward overview for the IPD dashboard."""

    total_wards: int
    total_beds: int
    occupied_beds: int
    available_beds: int
    active_admissions: int
    occupancy_rate: float
